src/app/util/context.h

```diff
@@ -0,0 +1,138 @@
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+
+#pragma once
+
+#include "config.h"
+
+#include <app/util/types_stub.h>
+#include <messaging/ExchangeDelegate.h>
+
+/**
+ *  @brief
+ *    Singleton handler for DataModel messages. Will be registered into ExchangeManager
+ */
+class DataModelHandler : public chip::ExchangeDelegate
+{
+public:
+    DataModelHandler() {}
+    virtual ~DataModelHandler() {}
+
+    // Not copyable or movable
+    DataModelHandler(const DataModelHandler &) = delete;
+    DataModelHandler & operator=(const DataModelHandler &) = delete;
+    DataModelHandler(DataModelHandler &&) = delete;
+    DataModelHandler& operator=(DataModelHandler &&) = delete;
+
+    /* ExchangeDelegate interface functions */
+    void OnMessageReceived(chip::ExchangeContext * ec, const chip::PacketHeader & packetHeader, uint32_t protocolId, uint8_t msgType, chip::System::PacketBuffer * payload) override;
+    void OnResponseTimeout(chip::ExchangeContext * ec) override;
+    void OnExchangeClosing(chip::ExchangeContext * ec) override;
+};
+
+/**
+ *  @brief
+ *    The ember transaction which is associated with a Exchange. The lifespan of the object is same as the ExchangeContext or slight
+ *    longer than the ExchangeContext. The initiator side should manually allocate the object, and bind it to a ExchangeContext
+ *    using ExchangeManager::NewContext. The acceptor side should manually allocate the object when first message received in the
+ *    registered UnsolicitedMessageHandler.
+ */
+class DataModelTransaction : public chip::ExchangeDelegate
+{
+public:
+    DataModelTransaction() {}
+    virtual ~DataModelTransaction() {}
+
+    // Not copyable or movable
+    DataModelTransaction(const DataModelTransaction &) = delete;
+    DataModelTransaction & operator=(const DataModelTransaction &) = delete;
+    DataModelTransaction(DataModelTransaction &&) = delete;
+    DataModelTransaction& operator=(DataModelTransaction &&) = delete;
+
+    /* ExchangeDelegate interface functions */
+    void OnMessageReceived(chip::ExchangeContext * ec, const chip::PacketHeader & packetHeader, uint32_t protocolId, uint8_t msgType, chip::System::PacketBuffer * payload) override;
+    void OnResponseTimeout(chip::ExchangeContext * ec) override;
+    void OnExchangeClosing(chip::ExchangeContext * ec) override;
+
+private:
+    // TODO: is there any member belong to the transaction ?
```

Comment:
Yes, I would expect the `ExchangeContext` here, no?  And then we don't need a separate `DataModelContext`....

The ownership (who frees what when) situation will, of course, need to get sorted.  It's really unclear so far for `ExchangeContext` how that's supposed to work...

Reply:
The `ExchangeContext` doesn't belong here. Because `ExchangeContext` will be always passed in as an augment.

If there is nothing in the `Transaction`, I can remove this class. `Transaction` should be dynamically allocated or use a object pool, it is better that we don't need it.

Reply:
> ExchangeContext will be always passed in as an augment.

When _receiving_ messages.  How do we represent sending outgoing exchange-initiating messages?

I'd basically like to have some representation of the concept of a "data model transaction".  Per spec, this happens to be synonymous with a "message layer exchange".  Is there an existing object that represents that thing?  My impression was that `ExchangeContext` was it, but if not, we need to figure out what that is.  If that _is_ it, then it can't quite represent a transaction, because it needs a delegate to handle receipt of messages (is that true?), so a transaction needs to encapsulate that delegate and the `ExchangeContext` somehow.

The things a "data model transaction" needs to support are:

1. Sending initial message when transaction is created.  This needs the `ExchangeContext` at first.
2. Receiving messages and responding to them.  This needs an `ExchangeDelegate`.
3. Controlling the lifetime of the exchange: closing it on "last message" send, possibly on timeouts, etc.  This is doable with just the delegate if the timeout bits are handled entirely in the message layer.  If the timeout logic lives in the interaction model layer, then it can't just use the delegate, afaict.